controllers/sync.go

```diff
@@ -393,60 +393,75 @@ type updateMeta struct {
 	componentNames []string
 }
 
-func canUpdateComponent(selector ytv1.UpdateSelector, component consts.ComponentType) bool {
-	switch selector {
-	case ytv1.UpdateSelectorNothing:
-		return false
-	case ytv1.UpdateSelectorMasterOnly:
-		return component == consts.MasterType
-	case ytv1.UpdateSelectorDataNodesOnly:
-		return component == consts.DataNodeType
-	case ytv1.UpdateSelectorTabletNodesOnly:
-		return component == consts.TabletNodeType
-	case ytv1.UpdateSelectorExecNodesOnly:
-		return component == consts.ExecNodeType
-	case ytv1.UpdateSelectorStatelessOnly:
-		switch component {
-		case consts.MasterType:
-			return false
-		case consts.DataNodeType:
-			return false
-		case consts.TabletNodeType:
-			return false
-		}
-		return true
-	case ytv1.UpdateSelectorEverything:
-		return true
-	default:
-		return false
+func getFlowFromComponent(component consts.ComponentType) ytv1.UpdateFlow {
+	if component == consts.MasterType {
+		return ytv1.UpdateFlowMaster
 	}
+	if component == consts.TabletNodeType {
+		return ytv1.UpdateFlowTabletNodes
+	}
+	if component == consts.DataNodeType || component == consts.ExecNodeType {
+		return ytv1.UpdateFlowFull
+	}
+	return ytv1.UpdateFlowStateless
+}
```

Comment:
I see this PR is raw yet, but just want to emphasize some things.
Obviously to add updateSelectors field is not challenging part of the issue, but to correctly update each component (or combination of components) selected by it, so all required pre and post steps were executed.

Currently we have some flows for component updates:
**master**
- PossibilityCheck
- WaitingForSafeModeEnabled
- WaitingForSnapshots
- WaitingForPodsRemoval
- WaitingForPodsCreation
- WaitingForMasterExitReadOnly
- WaitingForSafeModeDisabled

**tablet nodes**
- PossibilityCheck
- WaitingForTabletCellsSaving
- WaitingForTabletCellsRemovingStart
- WaitingForTabletCellsRemoved
- WaitingForPodsRemoval
- WaitingForPodsCreation
- WaitingForTabletCellsRecovery

**stateless**
- WaitingForPodsRemoval
- WaitingForPodsCreation
- WaitingForOpArchiveUpdatingPrepare
- WaitingForOpArchiveUpdate
- WaitingForQTStateUpdatingPrepare
- WaitingForQTStateUpdate

**everything**
- PossibilityCheck
- WaitingForSafeModeEnabled
- WaitingForTabletCellsSaving
- WaitingForTabletCellsRemovingStart
- WaitingForTabletCellsRemoved
- WaitingForSnapshots
- WaitingForPodsRemoval
- WaitingForPodsCreation
- WaitingForMasterExitReadOnly
- WaitingForTabletCellsRecovery
- WaitingForOpArchiveUpdatingPrepare
- WaitingForOpArchiveUpdate
- WaitingForQTStateUpdatingPrepare
- WaitingForQTStateUpdate
- WaitingForSafeModeDisabled

As you see `everything` includes (basically copy-pastes) parts `master`,`tnds`,`stateless` flows.
And 
if you want implement `UpdateSelectors = {component=scheduler}` you have to add flow for scheduler with `WaitingForOpArchiveUpdatingPrepare/WaitingForOpArchiveUpdate/WaitingForQTStateUpdatingPrepare/WaitingForQTStateUpdate`
if you want to implement `UpdateSelectors = {component=qt}` you have to add flow with `WaitingForQTStateUpdatingPrepare/WaitingForQTStateUpdate`
and so on.
So this predefined flows code is a dead end way if we want to achieve update granularity we want. So we should somehow concatenate workflows for particular components.

And i think the most simplest way here is to do this change with introducing Linear update (https://github.com/ytsaurus/ytsaurus-k8s-operator/issues/115).
I assume the flow in the code should look something like that:
```
# update flow
if discoveryType in toUpdateList: // updatable by selector and require update
	remove/create discovery pods

if httpProxy in updateList:
       remove/create hp pods
// tcp proxy
// rpc proxy
// data nodes
// exec nodes
// master cache

if tabletNodes in updateList:
     - PossibilityCheck
     - WaitingForTabletCellsSaving
     - WaitingForTabletCellsRemovingStart
     - WaitingForTabletCellsRemoved
     - WaitingForPodsRemoval
     - WaitingForPodsCreation
     - WaitingForTabletCellsRecovery

// ui
// ca
// yqla (but soon also will have pre-update jobs)
// sch with its flow
// qt with its flow
// qa with its flow
// strawberry
// master with its flow (updating in the end, because it should never be newer than other components)
```
Initializing flow is the same, but master in the beginning.

So basically I think we need here to:
1) resolve updateSelectors in a list of allowed components, match it with all component's statuses so we have a list of components we plan to update, sort it according to predefined order, choose first one we will update in this reconcillation loop, put it in `UpdateStatus.Components` 
2) check current `UpdateStatus.State` and `UpdateStatus.Conditions` and either call components.sync and requeue or set next State of this component flow and requeue.

I understand it is a bit vague overview, but maybe it helps.